esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,289 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+        if (jsonLayout.length())
+            accessInfo.setRecordTypeInfoJson(jsonLayout.str());
+    }
+    if (req.getReturnBinTypeInfo())
+    {
+        MemoryBuffer binLayout;
+        layoutBin.swapWith(binLayout);
+        if (binLayout.length())
+            accessInfo.setRecordTypeInfoBin(binLayout);
+    }
+}
+
+const char DFUFileIdSeparator = ':';
+
+bool CWsDfuEx::onDFUFileCreate(IEspContext &context, IEspDFUFileCreateRequest &req, IEspDFUFileCreateResponse &resp)
+{
+    try
+    {
+        IConstDFUFileAccessRequestBase &requests = req.getRequestBase();
+        SecAccessFlags accessType = translateToSecAccessFlags(requests.getAccessType());
+        if (SecAccess_None == accessType)
+        {
+            context.setAuthStatus(AUTH_STATUS_NOACCESS);
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate - Permission denied.");
+        }
+        context.ensureFeatureAccess(FEATURE_URL, accessType, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate: Permission denied.");
+
+        const char *fileName = requests.getName();
+        const char *clusterName = requests.getCluster();
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (isEmptyString(fileName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Name defined.");
+        if (isEmptyString(clusterName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Cluster defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No ECLRecordDefinition defined.");
+
+        StringBuffer groupName;
+        if (!createWsDFUCreateGroupName(req.getPartLocations(), groupName))
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in createWsDFUCreateGroupName().");
+
+        Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName);
+        ClusterType clusterType = clusterInfo->getPlatform();
+        if (clusterType == RoxieCluster)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Roxie cluster is not allowed: %s.", clusterName);
+
+        MemoryBuffer layoutBin;
+        exportRecordDefinitionBinaryType(recordDefinition, layoutBin);
+
+        StringBuffer userId;
+        Owned<IUserDescriptor> userDesc;
+        context.getUserID(userId);
+        if (!userId.isEmpty())
+        {
+            userDesc.setown(createUserDescriptor());
+            userDesc->set(userId.str(), context.queryPassword(), context.querySignature());
+        }
+
+        StringBuffer tempFileName = fileName;
+        tempFileName.append(".wsdfucreate.tmp");
+
+        StringBuffer jobId = requests.getJobId();
+        if (jobId.isEmpty())
+            jobId.appendf("Create %s on %s", tempFileName.str(), clusterName);
+        ESPLOG(LogMax, "%s", jobId.str());
+        Owned<IFileDescriptor> fileDesc = createDFUDescriptor(context, tempFileName.str(),
```

Comment:
createDFUDescriptor sounds like a utility function that should belong in dafdesc.cpp (if there isn't something similar already).

NB: I see 'context' isn't used in it.

Reply:
Yes, the context is not used now. It carries some ESP information which may be used later. For example, ESP client version. ESP may return different data based on the version in case some clients may not accept some data added by new version of ESP. IF it is moved to dafdesc.cpp, the context can be removed.

Reply:
I think it's of general use and not specific to WsDfu (or esp) - so worth moving.